### PR TITLE
Refactoring docker client. Image registry changes.

### PR DESCRIPTION
**Pull request by tushar-tushar**
Docker binary client will store a local copy of config and use it across operations, which is cleaned up after image build.
Build Context will store a map of image registries, this will support multiple pull registries.

### CODE REVIEW

File: workflow-controller/controller-service/src/main/java/io/hyscale/controller/invoker/ImageBuildComponentInvoker.java [hunk 1]
```diff
@@ -113,40 +128,53 @@ protected void doExecute(WorkflowContext context) throws HyscaleException {
...
+        }
+        return pullRegistries;
     }
 
     @Override
     protected void onError(WorkflowContext context, HyscaleException he) throws HyscaleException {
         WorkflowLogger.header(ControllerActivity.ERROR);
-        WorkflowLogger.error(ControllerActivity.CAUSE, he != null ? he.getMessage() : ImageBuilderErrorCodes.FAILED_TO_BUILD_AND_PUSH_IMAGE.getMessage());
-        context.addAttribute(WorkflowConstants.ERROR_MESSAGE, (he != null) ? he.getMessage() : ImageBuilderErrorCodes.FAILED_TO_BUILD_AND_PUSH_IMAGE.getMessage());
+        WorkflowLogger.error(ControllerActivity.CAUSE,
```
Comment (nagachandra-immadisetty) [inline workflow-controller/controller-service/src/main/java/io/hyscale/controller/invoker/ImageBuildComponentInvoker.java:174]: In case of structured o/p?

Comment (tushar-tushar) [reply workflow-controller/controller-service/src/main/java/io/hyscale/controller/invoker/ImageBuildComponentInvoker.java:174]: Haven't made any changes for that scenario, so it is being handled as it was earlier.

File: image-builder/builder-services/src/main/java/io/hyscale/builder/services/docker/HyscaleDockerClient.java [hunk 2]
```diff
@@ -31,7 +31,18 @@
  *
  */
 public interface HyscaleDockerClient {
-
+    
+    /**
+     * @return true if manual login is required for the client
+     */
+    boolean loginRequired();
```
Comment (nagachandra-immadisetty) [inline image-builder/builder-services/src/main/java/io/hyscale/builder/services/docker/HyscaleDockerClient.java:38]: change method name to isLoginRequired()

File: image-builder/builder-services/src/main/java/io/hyscale/builder/services/docker/HyscaleDockerClient.java [hunk 3]
```diff
@@ -31,7 +31,18 @@
...
+    
+    /**
+     * @return true if clean up temporary files are required
+     */
+    default boolean cleanUp(){
```
Comment (nagachandra-immadisetty) [inline image-builder/builder-services/src/main/java/io/hyscale/builder/services/docker/HyscaleDockerClient.java:43]: change method name to isCleanUpRequired()

File: image-builder/builder-services/src/main/java/io/hyscale/builder/services/docker/impl/DockerRESTClient.java [hunk 4]
```diff
@@ -339,14 +346,16 @@ public void onNext(PushResponseItem item) {
...
-        } catch (InterruptedException  e) {
+        } catch (InterruptedException e) {
             Thread.currentThread().interrupt();
-            handleOutput(buildContext.isVerbose(), pushActivity, Status.FAILED);
+            handleOutput(isVerbose, pushActivity, Status.FAILED);
             throw new HyscaleException(e, ImageBuilderErrorCodes.FAILED_TO_PUSH_IMAGE);
         }
-        handleOutput(buildContext.isVerbose(), pushActivity, Status.DONE);
+        handleOutput(isVerbose, pushActivity, Status.DONE);
+        String shaSum = sb.toString();
```
Comment (nagachandra-immadisetty) [inline image-builder/builder-services/src/main/java/io/hyscale/builder/services/docker/impl/DockerRESTClient.java:357]: 1.Remove this shaSum String variable. 
2. Rename sb to shaSum.
3. Update return statement as **return StringUtils.isNotBlank(shaSum) ? shaSum.toString() : null;**

File: image-builder/builder-services/src/main/java/io/hyscale/builder/services/docker/impl/DockerRESTClient.java [hunk 5]
```diff
@@ -374,13 +383,22 @@ private void handleOutput(boolean isVerbose, ActivityContext activityContext, St
...
     private AuthConfig getAuthConfig(ImageRegistry imageRegistry) {
         if (imageRegistry == null) {
             return null;
         }
         AuthConfig authConfig = new AuthConfig();
         authConfig.withRegistryAddress(imageRegistry.getUrl());
-        String decodedAuth = new String(Base64.getDecoder().decode(imageRegistry.getToken()));
+        String decodedAuth = EncodeDecodeUtil.decode(imageRegistry.getToken());
         int delimiter = decodedAuth.indexOf(':');
         if (delimiter > 0) {
```
Comment (nagachandra-immadisetty) [inline image-builder/builder-services/src/main/java/io/hyscale/builder/services/docker/impl/DockerRESTClient.java:403]: Can we simplify this into a separate method in EncodeDecodeUtil?
like getDecodedUserName()
what say?

**Review by nagachandra-immadisetty — CHANGES_REQUESTED** [review]
Let me know any clarification required over the comments.

**Review by nagachandra-immadisetty — APPROVED** [review]
Looks Good to me.

**Review by harijvs6 — APPROVED** [review]

**Review by harijvs6 — APPROVED** [review]

**Pull request merged**